lamindb/dev/db/_add.py

```diff
@@ -81,28 +81,84 @@ def add(  # type: ignore
             return results
         else:
             records = [model(**fields)]
-    for record in records:
-        if isinstance(record, DObject) and hasattr(record, "_local_filepath"):
-            upload_data_object(record, use_fsspec=use_fsspec)
+
     if session is None:  # assume global session
         session = settings.instance.session()
         settings.instance._cloud_sqlite_locker.lock()
         close = True
     else:
         close = False
+
+    added_records = []
+    raise_error = False
     for record in records:
-        session.add(record)
-    session.commit()
```

Comment:
Most calls of `ln.add()` will affect ORMs that are non-DObject, and hence, to not have a corresponding upload of a (often large) file associate to them

We should retain the performance and safety benefits from committing all records in a list of records at the same time.

Say, we add 30k gene records: We'd not want to run 30k db transactions for these very light-weight operations

Hence, we need clean logic that separates ingesting a list of dobjects from a list of non-dobjects

Reply:
Understood. I've updated the logic and implemented single-DB-transaction for all metadata ingestion cases. In case of failure at upload, I simply create a transaction to delete the relevant metadata record.